src/net/client/xfr.rs

```diff
@@ -0,0 +1,1147 @@
+
+        let iter = XfrEventIterator {
+            req,
+            state,
+            records,
+        };
+
+        Ok(iter)
+    }
+
+    /// Process a subsequent XFR response message.
+    ///
+    /// Revives the given iterator with new data based on the given response.
+    pub async fn process_next_answer<'a>(
+        it: &mut XfrEventIterator<'a>,
+        resp: &'a Message<Bytes>,
+    ) -> Result<(), Error> {
+        // Verify that the given iterator is in the expected state
+        match it.state {
+            State::AwaitingAnswer { .. } => {
+                // Verify that the given iterator is exhausted
+                if it.records.next().is_some() {
+                    return Err(Error::AnswerNotFullyProcessed);
+                }
+
+                let _ = Self::check_is_xfr_answer(it.req, resp).await?;
+                let answer = resp.answer().map_err(Error::ParseError)?;
+                it.records = answer.into_records().flatten();
+
+                Ok(())
+            }
+
+            State::TransferComplete => {
+                // We already finished processing an XFR response sequence. We
+                // don't expect there to be any more messages to process!.
+                Err(Error::Malformed)
+            }
+
+            State::TransferFailed => {
+                // We had to terminate processing of the XFR response sequence
+                // due to a problem with the received data, so we don't expect
+                // to be invoked again with another response message!
+                Err(Error::Terminated)
+            }
+        }
+    }
+
+    /// Check if an XFR response header is valid.
+    ///
+    /// Enforce the rules defined in 2. AXFR Messages of RFC 5936. See:
+    /// https://www.rfc-editor.org/rfc/rfc5936.html#section-2
+    ///
+    /// Takes a request as well as a response as the response is checked to
+    /// see if it is in reply to the given request.
+    ///
+    /// Returns Ok on success, Err otherwise. On success the type of XFR that
+    /// was determined is returned as well as the answer section from the XFR
+    /// response.
+    async fn check_is_xfr_answer(
+        req: &Message<Bytes>,
+        resp: &Message<Bytes>,
+    ) -> Result<XfrType, CheckError> {
+        // Check the request.
+        let req_header = req.header();
+        let req_counts = req.header_counts();
+
+        if req.is_error()
+            || req_header.qr()
+            || req_counts.qdcount() != 1
+            || req_counts.ancount() != 0
+            || req_header.opcode() != Opcode::QUERY
+        {
+            return Err(CheckError::NotValidXfrQuery);
+        }
+
+        let Some(qtype) = req.qtype() else {
+            return Err(CheckError::NotValidXfrResponse);
+        };
+
+        let xfr_type = match qtype {
+            Rtype::AXFR => XfrType::Axfr,
+            Rtype::IXFR => XfrType::Ixfr,
+            _ => return Err(CheckError::NotValidXfrResponse),
+        };
```

Comment:
This doesn't belong in this function. It should check the reply not the request. Maybe another function should check the request.

Reply:
The current code still has this block but not for the purpose of checking the request but for determining if it is AXFR or IXFR.